src/EditorFeatures/CSharpTest/AddMissingImports/CSharpAddMissingImportsFeatureServiceTests.cs

```diff
@@ -0,0 +1,151 @@
+}
+";
+
+            var expected = @"
+using A;
+
+class C
+{
+    public D Foo { get; }
+}
+
+namespace A
+{
+    public class D { }
+}
+";
+
+            await AssertDocumentChangedAsync(code, expected).ConfigureAwait(false);
+        }
+
+        [Fact]
+        public async Task AddMissingImports_DocumentUnchanged_SpanContainsAmbiguousImports()
+        {
+            var code = @"
+class C
+{
+    public [|D|] Foo { get; }
+}
+
+namespace A
+{
+    public class D { }
+}
+
+namespace B
+{
+    public class D { }
+}
+";
+
+            await AssertDocumentUnchangedAsync(code).ConfigureAwait(false);
+        }
+
+        private async Task AssertDocumentUnchangedAsync(string initialMarkup)
+        {
+            using (var workspace = TestWorkspace.CreateCSharp(initialMarkup))
+            {
+                var diagnosticAnalyzerService = InitializeDiagnosticAnalyzerService(workspace);
+
+                var addMissingImportsService = new CSharpAddMissingImportsFeatureService(diagnosticAnalyzerService);
+
+                var hostDocument = workspace.Documents.First();
+                var documentId = hostDocument.Id;
+                var textSpan = hostDocument.SelectedSpans.First();
+
+                var document = workspace.CurrentSolution.GetDocument(documentId);
+
+                var newProject = await addMissingImportsService.AddMissingImportsAsync(document, textSpan, CancellationToken.None).ConfigureAwait(false);
+                var newDocument = newProject.GetDocument(documentId);
+
+                Assert.Equal(document, newDocument);
+            }
+        }
+
+        private async Task AssertDocumentChangedAsync(string initialMarkup, string expectedMarkup)
+        {
+            using (var workspace = TestWorkspace.CreateCSharp(initialMarkup))
+            {
+                var diagnosticAnalyzerService = InitializeDiagnosticAnalyzerService(workspace);
+
+                var addMissingImportsService = new CSharpAddMissingImportsFeatureService(diagnosticAnalyzerService);
+
+                var hostDocument = workspace.Documents.First();
+                var documentId = hostDocument.Id;
+                var textSpan = hostDocument.SelectedSpans.First();
+
+                var document = workspace.CurrentSolution.GetDocument(documentId);
+
+                var newProject = await addMissingImportsService.AddMissingImportsAsync(document, textSpan, CancellationToken.None).ConfigureAwait(false);
+                var newDocument = newProject.GetDocument(documentId);
+
+                Assert.NotEqual(document, newDocument);
+
+                var text = await newDocument.GetTextAsync().ConfigureAwait(false);
+
+                Assert.Equal(expectedMarkup, text.ToString());
+            }
+        }
+
+        private IDiagnosticAnalyzerService InitializeDiagnosticAnalyzerService(Workspace workspace)
+        {
+            var diagnosticAnalyzer = DiagnosticExtensions.GetCompilerDiagnosticAnalyzer(LanguageName);
```

Comment:
this works but I think @sharwell probably want to get the diagnostic service from the MEF with some package provider.. not sure what his recommendation is, but he wants test to not use any kind of mock but get everything from MEF.